### DISCUSSION TITLE
Question about multilingual font support

### DISCUSSION THREAD

**Discussion by namsancow** [iTwin/itwinjs-core#1685] [Q&A]
The 2D DGN files including Korean fonts are well displayed in MicroStation and BentleyView.
However, the published BIM file from this does have the exact information about Korean fonts.
In Desktop Starter, all Korean fonts disappear or are replaced by an empty square.
Is there any special option for multilingual font support ?

**Comment by pmconne** [comment]
@jffmarker

**Reply by jffmarker** [reply]
There is no special configuration, but I can think of two potential issues off-hand. Perhaps the machine you're converting on (i.e. is this in the cloud or your local machine?) doesn't have the font specified in the DGN. Or, the DGN specifies a font that doesn't contain Korean characters, and you're relying on Windows-specific glyph-substitution ("font linking") that only MicroStation supports (there is no cross-platform equivalent for BIM to support)... do either of these situations seem likely?

**Reply by namsancow** [reply]
Dear jffmaker and pmconne

Thank you for your kind response.
According to your advice, I checked the setting option related in fonts of DGN file
However, I could not find the special issue about fonts setting.

I guess the bim file published by "iTwin SnapShot" does not contain Korean True Type Fonts.

I can not find Korean True Type Fonts in SQL DB of bim file. 
If I don't understand something, please let me know.

Best regards

dgn_fonts in bim file on SQLite (left) vs BentleyView (Right)
![KoreanFont06](https://user-images.githubusercontent.com/22971147/123216896-fa3fc200-d504-11eb-974a-ba98bc8647a1.jpg)

Korean fonts breaking on iModel Desktop-Starter 
![KoreanFont06 imodeljs output](https://user-images.githubusercontent.com/22971147/123216914-ff9d0c80-d504-11eb-9c5f-c48470c3623b.jpg)

**Reply by jffmarker** [reply]
So you're using the iTwin Snapshot generator to convert the DGN file on your own computer? Then it seems unlikely to be a missing font problem. And given the Korean name of the font, it seems unlikely to be a font linking problem.

The dgn_Font table is just an ID mapping (with optional metadata). Embedded font *data* are stored in be_Prop. To see which fonts were successfully embedded in the BIM file, you can run:

```sql
SELECT  DISTINCT json_extract(value, '$.familyName')
FROM    be_Prop, json_each(StrData)
WHERE   Namespace = 'dgn_Font'
```

Was the Korean font (sorry can't copy/paste the name from an image) successfully embedded in your snapshot?

**Reply by namsancow** [reply]
Dear jffmarker

I checked "be_Prop" in bim file according to your explanation. 
I used "Arial" font to express Korean Text "안녕하세요" instead of "굴림" or "맑은고딕" korean font.
In "be_Prop", Arial font is saved normally as following pictures.
However, the Korean font breaking keeps happening.
Is there any option to control font setting in imodel.js ?

Best regards

English and Korean text display with Arial Font  in Bentley View
![Arial Font English Korean Bentley View](https://user-images.githubusercontent.com/22971147/123579439-d9cc7c00-d812-11eb-92dc-76d59cc8cb0e.jpg)

Arial Font saved in  SQLite be_Prop
![Arial Font English Korean SQLite be_Prop](https://user-images.githubusercontent.com/22971147/123579476-eb158880-d812-11eb-88bb-0a03e25c698a.jpg)

English and Korean text display with Arial Font  in Desktop Starter
![Arial Font English Korean Desktop_Starter](https://user-images.githubusercontent.com/22971147/123579714-6119ef80-d813-11eb-98a2-5b254dc136cf.jpg)

**Reply by namsancow** [reply]
Dear jeffmarker

I found the mismatch between MstnFontConfig.xml and be_Prop & dgn_font of bim file. 
In MstnFontConfig.xml, "GulimChe" is default font as follows;

        <LanguageSetting Name="Korean">
            <DefaultRscFont>Gothic_single</DefaultRscFont>
            <DefaultTrueTypeFont>GulimChe</DefaultTrueTypeFont>
            <DefaultShxFont>kor</DefaultShxFont>
            <DefaultShxBigFont>whgtxt,korbig</DefaultShxBigFont>
            <FontLinking>Enabled</FontLinking>
            <GlyphShaping>Disabled</GlyphShaping>
        </LanguageSetting>

If the 2D file with Korean fonts is opened, All font are changed as "GulimChe".
However, after being published as bim file of imodel 2.0, "GulimChe" is changed as "LastResortTrueTypeFont".
And all Korean fonts are breaking in Desktop Starter.
Is there embedded converting font list in "iTwin SnapShot" ?

Best regards

**Reply by jffmarker** [reply]
Arial does not contain glyphs for Korean characters, so seeing missing glyph boxes in that scenario is expected. iModel.js does not support font linking, which is how Arial *appears* to work in MicroStation (glyphs are getting substituted on-the-fly based on a Microsoft Windows library).

If GulimChe is installed on your system, and the DGN elements use that font for Korean characters, then the iTwin Snapshot tool and Desktop Starter should work (the Connector will embed the font, and the viewer will show it). N.B. be sure to delete any *.Tiles files next to the .bim files between creating snapshots to ensure the display is updated between conversions.

MicroStation (left) vs. Desktop Starter (right) with font linking disabled:
![image](https://user-images.githubusercontent.com/8006049/123634654-b7be0400-d7e8-11eb-9e2b-48963addbe58.png)

I see entries in both be_Prop and dgn_Font for both Arial and GulimChe.
